src/main/python/deep_qa/layers/attention/weighted_sum.py

```diff
@@ -0,0 +1,91 @@
+    My attention "vector" must then have at least those dimensions, and could have more.  So I
+    could have an attention over words per query, with shape (batch_size, num_queries, num_words),
+    or I could have an attention over query words for every document in some list, with shape
+    (batch_size, num_documents, num_queries, num_words).  Both of these cases are fine.  In the
+    first case, the returned tensor will have shape (batch_size, num_queries, embedding_dim), and
+    in the second case, it will have shape (batch_size, num_documents, num_queries, embedding_dim).
+    But you _can't_ have an attention "vector" that does not include all of the queries, so shape
+    (batch_size, num_words) is not allowed - you haven't specified how to handle that dimension in
+    the "matrix", so we can't do anything with this input.
+    '''
+    def __init__(self, **kwargs):
+        self.supports_masking = True
+        super(WeightedSum, self).__init__(**kwargs)
+
+    @overrides
+    def compute_mask(self, inputs, mask=None):
+        # pylint: disable=unused-argument
+        # We don't need to worry about a mask after we've summed over the rows of the matrix.
+        # You might actually still need a mask if you used a higher-order tensor, but probably the
+        # right place to handle that is with careful use of TimeDistributed.  Or submit a PR.
+        return None
+
+    @overrides
+    def get_output_shape_for(self, input_shapes):
+        matrix_shape, attention_shape = input_shapes
+        return attention_shape[:-1] + matrix_shape[-1:]
+
+    @overrides
+    def call(self, inputs, mask=None):
+        matrix, attention_vector = inputs
+        matrix = self._expand_matrix_if_necessary(matrix, attention_vector)
+        matrix_mask = mask[0]
+        if matrix_mask is not None:
+            matrix_mask = K.expand_dims(matrix_mask, dim=-1)
+            matrix_mask = self._expand_matrix_if_necessary(matrix_mask, attention_vector)
+            matrix = switch(matrix_mask, matrix, K.zeros_like(matrix))
+        return K.sum(K.expand_dims(attention_vector, dim=-1) * matrix, -2)
+
+    @staticmethod
+    def _expand_matrix_if_necessary(matrix, attention_vector):
+        try:
+            matrix_shape = K.int_shape(matrix)[:-1]  # taking off the embedding dimension here.
+        except Exception:  # pylint: disable=broad-except
+            if K.backend() == 'theano':
+                raise RuntimeError("Theano backend doesn't support K.int_shape very well - use "
+                                   "tensorflow instead for this model.")
+            else:
+                raise
+        attention_shape = K.int_shape(attention_vector)
+        if matrix_shape != attention_shape:
+            # We'll take care of the batch size first.  After this, the matrix_shape should match
+            # the end of the attention_shape exactly.
+            assert matrix_shape[0] == attention_shape[0], "somehow batch sizes don't match"
```

Comment:
i'm actually not too sure, it frankly looked a bit like a debug message that you used in testing to me when I first saw it.